### PR TITLE
Issue #401 DateTime min and max value serialization

### PR DESCRIPTION
**Pull request by CrazyViper**

### CODE REVIEW

File: OData/src/System.Web.OData/OData/Formatter/Serialization/ODataPrimitiveSerializer.cs [hunk 1]
```diff
@@ -144,14 +144,26 @@ internal static object ConvertUnsupportedPrimitives(object value)
 
                     case TypeCode.DateTime:
                         DateTime dateTime = (DateTime)value;
-                        TimeZoneInfo timeZone = TimeZoneInfoHelper.TimeZone;
-                        if (dateTime.Kind == DateTimeKind.Utc || dateTime.Kind == DateTimeKind.Local)
+
+                        if (dateTime == DateTime.MinValue)
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Formatter/Serialization/ODataPrimitiveSerializer.cs]: We need to check if dateTime >= DateTime.MinValue + timeZone

File: OData/src/System.Web.OData/OData/Formatter/Serialization/ODataPrimitiveSerializer.cs [hunk 2]
```diff
@@ -144,14 +144,26 @@ internal static object ConvertUnsupportedPrimitives(object value)
...
                         {
-                            return new DateTimeOffset(dateTime.ToUniversalTime()).ToOffset(timeZone.BaseUtcOffset);
+                            return DateTimeOffset.MinValue;
+                        }
+                        else if (dateTime == DateTime.MaxValue)
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Formatter/Serialization/ODataPrimitiveSerializer.cs]: Similar to MaxValue

File: OData/test/UnitTest/System.Web.OData.Test/OData/Formatter/Serialization/ODataPrimitiveSerializerTests.cs [hunk 3]
```diff
@@ -164,6 +164,42 @@ public void CreateODataValue_ReturnsDateTimeOffset_ForDateTime_ByDefault()
             Assert.Equal(new DateTimeOffset(dt), primitiveValue.Value);
         }
 
+        [Fact]
```
Comment (congysu) [inline OData/test/UnitTest/System.Web.OData.Test/OData/Formatter/Serialization/ODataPrimitiveSerializerTests.cs]: Add another 4 test cases:
MinValue, with a TimeZone > 0
MinValue, with a TimeZone < 0
MaxValue, with a TimeZone > 0
MaxValue, with a TimeZone < 0

**Comment by CrazyViper** [comment]
> We need to check if dateTime >= DateTime.MinValue + timeZone
Maybe you wanted to say if dateTime >= DateTime.MinValue + timeZone.BaseUtcOffset? If so, you will get ArgumentOutOfRangeException when timeZone.BaseUtcOffset is negative.

MinValue has always Kind = Unspecified (for DateTimeOffset property Offset set to zero). I believe we have to transfer it to client with zero offset.

File: OData/src/System.Web.OData/OData/Formatter/Serialization/ODataPrimitiveSerializer.cs (hunk 1 above)
Comment (CrazyViper) [inline OData/src/System.Web.OData/OData/Formatter/Serialization/ODataPrimitiveSerializer.cs]: Maybe you wanted to say if dateTime >= DateTime.MinValue + timeZone.BaseUtcOffset? 
If so, you will get ArgumentOutOfRangeException when timeZone.BaseUtcOffset is negative.

MinValue has always Kind = Unspecified (for DateTimeOffset property Offset set to zero). I believe we have to transfer it to client with zero offset.

Comment (congysu) [inline OData/src/System.Web.OData/OData/Formatter/Serialization/ODataPrimitiveSerializer.cs]: For example, if `dateTime` = '0001-01-01 01:00:00', but the `timeZone` is GMT+8; then the line 165, 166 might throw out of range exception.

**Comment by xuzhg** [comment]
@CrazyViper Sorry for response later. 

I try to take the timezone consideration. So I modify your PR as blow. Would you please take a look and try it out?

```C#

DateTime dateTime = (DateTime) value;

TimeZoneInfo timeZone = TimeZoneInfoHelper.TimeZone;
if (timeZone.BaseUtcOffset.Hours >= 0)
{
    if (dateTime <= DateTime.MinValue + timeZone.BaseUtcOffset)
    {
        return DateTimeOffset.MinValue;
    }
}
else
{
    if (dateTime >= DateTime.MaxValue + timeZone.BaseUtcOffset)
    {
        return DateTimeOffset.MaxValue;
    }
}

if (dateTime.Kind == DateTimeKind.Local)
{
    TimeZoneInfo localTimeZoneInfo = TimeZoneInfo.Local;
    if (localTimeZoneInfo.BaseUtcOffset.Hours < 0)
    {
        if (dateTime >= DateTime.MaxValue + localTimeZoneInfo.BaseUtcOffset)
        {
            return DateTimeOffset.MaxValue;
        }
    }
    else
    {
        if (dateTime <= DateTime.MinValue + localTimeZoneInfo.BaseUtcOffset)
        {
            return DateTimeOffset.MinValue;
        }
    }

    return TimeZoneInfo.ConvertTime(new DateTimeOffset(dateTime), timeZone);
}

if (dateTime.Kind == DateTimeKind.Utc)
{
    return TimeZoneInfo.ConvertTime(new DateTimeOffset(dateTime), timeZone);
}

return new DateTimeOffset(dateTime, timeZone.GetUtcOffset(dateTime));
```

**Comment by xuzhg** [comment]
@CrazyViper 

I create a console application to test. In which i add a function with the above codes as function body.

```C#
private static DateTimeOffset ConvertDateTimeToTimeOffset(DateTime dateTime, TimeZoneInfo timeZone)
{
    if (timeZone.BaseUtcOffset.Hours >= 0)
    {
        if (dateTime <= DateTime.MinValue + timeZone.BaseUtcOffset)
        {
            return DateTimeOffset.MinValue;
        }
    }
    else
    {
        if (dateTime >= DateTime.MaxValue + timeZone.BaseUtcOffset)
        {
            return DateTimeOffset.MaxValue;
        }
    }

    if (dateTime.Kind == DateTimeKind.Local)
    {
        TimeZoneInfo localTimeZoneInfo = TimeZoneInfo.Local;
        if (localTimeZoneInfo.BaseUtcOffset.Hours < 0)
        {
            if (dateTime >= DateTime.MaxValue + localTimeZoneInfo.BaseUtcOffset)
            {
                return DateTimeOffset.MaxValue;
            }
        }
        else
        {
            if (dateTime <= DateTime.MinValue + localTimeZoneInfo.BaseUtcOffset)
            {
                return DateTimeOffset.MinValue;
            }
        }

        return TimeZoneInfo.ConvertTime(new DateTimeOffset(dateTime), timeZone);
    }

    if (dateTime.Kind == DateTimeKind.Utc)
    {
        return TimeZoneInfo.ConvertTime(new DateTimeOffset(dateTime), timeZone);
    }

    return new DateTimeOffset(dateTime, timeZone.GetUtcOffset(dateTime));
}
```


Besides, I use the following value to test:

```C#
DateTime[] dateTimes =
      {
                DateTime.MinValue,
                DateTime.MinValue.AddMinutes(2),
                DateTime.MinValue.AddHours(2),
                DateTime.MinValue.AddDays(2),
                DateTime.MinValue.AddYears(2),

                DateTime.MaxValue,
                DateTime.MaxValue.AddMinutes(-2),
                DateTime.MaxValue.AddHours(-2),
                DateTime.MaxValue.AddDays(-2),
                DateTime.MaxValue.AddYears(-2),

                new DateTime(2015, 7, 9, 1, 2, 3, DateTimeKind.Unspecified),
                new DateTime(2015, 7, 9, 1, 2, 3, DateTimeKind.Utc),
                new DateTime(2015, 7, 9, 1, 2, 3, DateTimeKind.Local),

                new DateTime(1, 1, 1, 1, 1, 1),
                new DateTime(1, 1, 1, 1, 1, 1, DateTimeKind.Utc),
                new DateTime(1, 1, 1, 1, 1, 1, DateTimeKind.Local),

                new DateTime(9999, 12, 31, 23, 1, 1),
                new DateTime(9999, 12, 31, 23, 1, 1, DateTimeKind.Utc),
                new DateTime(9999, 12, 31, 23, 1, 1,  DateTimeKind.Local),
            };

TimeZoneInfo utc = TimeZoneInfo.Utc;
TimeZoneInfo timeZone = TimeZoneInfo.FindSystemTimeZoneById("Pacific Standard Time");
TimeZoneInfo chinaStandard = TimeZoneInfo.FindSystemTimeZoneById("China Standard Time");

DateTimeOffset dto;
foreach (TimeZoneInfo tz in new[] { timeZone, utc, chinaStandard })
{
        Console.WriteLine("\n[TimeZone]: " + tz);
        foreach (var dateTime in dateTimes)
        {
               dto = ConvertDateTimeToTimeOffset(dateTime, tz);
               Console.WriteLine(" |--[" +dateTime + ", " + dateTime.Kind + "] => " + dto);
        }
}
```

And I get the following output:
```C#
[TimeZone]: (UTC-08:00) Pacific Time (US & Canada)
 |--[1/1/0001 12:00:00 AM, Unspecified] => 1/1/0001 12:00:00 AM -08:00
 |--[1/1/0001 12:02:00 AM, Unspecified] => 1/1/0001 12:02:00 AM -08:00
 |--[1/1/0001 2:00:00 AM, Unspecified] => 1/1/0001 2:00:00 AM -08:00
 |--[1/3/0001 12:00:00 AM, Unspecified] => 1/3/0001 12:00:00 AM -08:00
 |--[1/1/0003 12:00:00 AM, Unspecified] => 1/1/0003 12:00:00 AM -08:00
 |--[12/31/9999 11:59:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:57:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 9:59:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/29/9999 11:59:59 PM, Unspecified] => 12/29/9999 11:59:59 PM -08:00
 |--[12/31/9997 11:59:59 PM, Unspecified] => 12/31/9997 11:59:59 PM -08:00
 |--[7/9/2015 1:02:03 AM, Unspecified] => 7/9/2015 1:02:03 AM -07:00
 |--[7/9/2015 1:02:03 AM, Utc] => 7/8/2015 6:02:03 PM -07:00
 |--[7/9/2015 1:02:03 AM, Local] => 7/8/2015 10:02:03 AM -07:00
 |--[1/1/0001 1:01:01 AM, Unspecified] => 1/1/0001 1:01:01 AM -08:00
 |--[1/1/0001 1:01:01 AM, Utc] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 1:01:01 AM, Local] => 1/1/0001 12:00:00 AM +00:00
 |--[12/31/9999 11:01:01 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:01:01 PM, Utc] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 11:59:59 PM +00:00

[TimeZone]: UTC
 |--[1/1/0001 12:00:00 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 12:02:00 AM, Unspecified] => 1/1/0001 12:02:00 AM +00:00
 |--[1/1/0001 2:00:00 AM, Unspecified] => 1/1/0001 2:00:00 AM +00:00
 |--[1/3/0001 12:00:00 AM, Unspecified] => 1/3/0001 12:00:00 AM +00:00
 |--[1/1/0003 12:00:00 AM, Unspecified] => 1/1/0003 12:00:00 AM +00:00
 |--[12/31/9999 11:59:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:57:59 PM, Unspecified] => 12/31/9999 11:57:59 PM +00:00
 |--[12/31/9999 9:59:59 PM, Unspecified] => 12/31/9999 9:59:59 PM +00:00
 |--[12/29/9999 11:59:59 PM, Unspecified] => 12/29/9999 11:59:59 PM +00:00
 |--[12/31/9997 11:59:59 PM, Unspecified] => 12/31/9997 11:59:59 PM +00:00
 |--[7/9/2015 1:02:03 AM, Unspecified] => 7/9/2015 1:02:03 AM +00:00
 |--[7/9/2015 1:02:03 AM, Utc] => 7/9/2015 1:02:03 AM +00:00
 |--[7/9/2015 1:02:03 AM, Local] => 7/8/2015 5:02:03 PM +00:00
 |--[1/1/0001 1:01:01 AM, Unspecified] => 1/1/0001 1:01:01 AM +00:00
 |--[1/1/0001 1:01:01 AM, Utc] => 1/1/0001 1:01:01 AM +00:00
 |--[1/1/0001 1:01:01 AM, Local] => 1/1/0001 12:00:00 AM +00:00
 |--[12/31/9999 11:01:01 PM, Unspecified] => 12/31/9999 11:01:01 PM +00:00
 |--[12/31/9999 11:01:01 PM, Utc] => 12/31/9999 11:01:01 PM +00:00
 |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 3:01:01 PM +00:00

[TimeZone]: (UTC+08:00) Beijing, Chongqing, Hong Kong, Urumqi
 |--[1/1/0001 12:00:00 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 12:02:00 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 2:00:00 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/3/0001 12:00:00 AM, Unspecified] => 1/3/0001 12:00:00 AM +08:00
 |--[1/1/0003 12:00:00 AM, Unspecified] => 1/1/0003 12:00:00 AM +08:00
 |--[12/31/9999 11:59:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +08:00
 |--[12/31/9999 11:57:59 PM, Unspecified] => 12/31/9999 11:57:59 PM +08:00
 |--[12/31/9999 9:59:59 PM, Unspecified] => 12/31/9999 9:59:59 PM +08:00
 |--[12/29/9999 11:59:59 PM, Unspecified] => 12/29/9999 11:59:59 PM +08:00
 |--[12/31/9997 11:59:59 PM, Unspecified] => 12/31/9997 11:59:59 PM +08:00
 |--[7/9/2015 1:02:03 AM, Unspecified] => 7/9/2015 1:02:03 AM +08:00
 |--[7/9/2015 1:02:03 AM, Utc] => 7/9/2015 9:02:03 AM +08:00
 |--[7/9/2015 1:02:03 AM, Local] => 7/9/2015 1:02:03 AM +08:00
 |--[1/1/0001 1:01:01 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 1:01:01 AM, Utc] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 1:01:01 AM, Local] => 1/1/0001 12:00:00 AM +00:00
 |--[12/31/9999 11:01:01 PM, Unspecified] => 12/31/9999 11:01:01 PM +08:00
 |--[12/31/9999 11:01:01 PM, Utc] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 11:01:01 PM +08:00
```

**Comment by CrazyViper** [comment]
@xuzhg Thanks for these modifications. I've tried to do something like this, but I still have the problem with local TZ. It seems that somewhere in CLR it uses local computer TimeZine to convert to datetimeoffset. And the results always different at the boundaries with different TZ. My local TZ is the UTC +3. I've got the following output with your test console app:

```
[TimeZone]: (UTC-08:00) Pacific Time (US & Canada)
 |--[1/1/0001 12:00:00 AM, Unspecified] => 1/1/0001 12:00:00 AM -08:00
 |--[1/1/0001 12:02:00 AM, Unspecified] => 1/1/0001 12:02:00 AM -08:00
 |--[1/1/0001 2:00:00 AM, Unspecified] => 1/1/0001 2:00:00 AM -08:00
 |--[1/3/0001 12:00:00 AM, Unspecified] => 1/3/0001 12:00:00 AM -08:00
 |--[1/1/0003 12:00:00 AM, Unspecified] => 1/1/0003 12:00:00 AM -08:00
 |--[12/31/9999 11:59:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:57:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 9:59:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/29/9999 11:59:59 PM, Unspecified] => 12/29/9999 11:59:59 PM -08:00
 |--[12/31/9997 11:59:59 PM, Unspecified] => 12/31/9997 11:59:59 PM -08:00
 |--[7/9/2015 1:02:03 AM, Unspecified] => 7/9/2015 1:02:03 AM -07:00
 |--[7/9/2015 1:02:03 AM, Utc] => 7/8/2015 6:02:03 PM -07:00
 |--[7/9/2015 1:02:03 AM, Local] => 7/8/2015 3:02:03 PM -07:00
 |--[1/1/0001 1:01:01 AM, Unspecified] => 1/1/0001 1:01:01 AM -08:00
 |--[1/1/0001 1:01:01 AM, Utc] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 1:01:01 AM, Local] => 1/1/0001 12:00:00 AM +00:00
 |--[12/31/9999 11:01:01 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:01:01 PM, Utc] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 11:59:59 PM +00:00

[TimeZone]: UTC
 |--[1/1/0001 12:00:00 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 12:02:00 AM, Unspecified] => 1/1/0001 12:02:00 AM +00:00
 |--[1/1/0001 2:00:00 AM, Unspecified] => 1/1/0001 2:00:00 AM +00:00
 |--[1/3/0001 12:00:00 AM, Unspecified] => 1/3/0001 12:00:00 AM +00:00
 |--[1/1/0003 12:00:00 AM, Unspecified] => 1/1/0003 12:00:00 AM +00:00
 |--[12/31/9999 11:59:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:57:59 PM, Unspecified] => 12/31/9999 11:57:59 PM +00:00
 |--[12/31/9999 9:59:59 PM, Unspecified] => 12/31/9999 9:59:59 PM +00:00
 |--[12/29/9999 11:59:59 PM, Unspecified] => 12/29/9999 11:59:59 PM +00:00
 |--[12/31/9997 11:59:59 PM, Unspecified] => 12/31/9997 11:59:59 PM +00:00
 |--[7/9/2015 1:02:03 AM, Unspecified] => 7/9/2015 1:02:03 AM +00:00
 |--[7/9/2015 1:02:03 AM, Utc] => 7/9/2015 1:02:03 AM +00:00
 |--[7/9/2015 1:02:03 AM, Local] => 7/8/2015 10:02:03 PM +00:00
 |--[1/1/0001 1:01:01 AM, Unspecified] => 1/1/0001 1:01:01 AM +00:00
 |--[1/1/0001 1:01:01 AM, Utc] => 1/1/0001 1:01:01 AM +00:00
 |--[1/1/0001 1:01:01 AM, Local] => 1/1/0001 12:00:00 AM +00:00
 |--[12/31/9999 11:01:01 PM, Unspecified] => 12/31/9999 11:01:01 PM +00:00
 |--[12/31/9999 11:01:01 PM, Utc] => 12/31/9999 11:01:01 PM +00:00
 |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 8:01:01 PM +00:00

[TimeZone]: (UTC+08:00) Beijing, Chongqing, Hong Kong, Urumqi
 |--[1/1/0001 12:00:00 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 12:02:00 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 2:00:00 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/3/0001 12:00:00 AM, Unspecified] => 1/3/0001 12:00:00 AM +08:00
 |--[1/1/0003 12:00:00 AM, Unspecified] => 1/1/0003 12:00:00 AM +08:00
 |--[12/31/9999 11:59:59 PM, Unspecified] => 12/31/9999 11:59:59 PM +08:00
 |--[12/31/9999 11:57:59 PM, Unspecified] => 12/31/9999 11:57:59 PM +08:00
 |--[12/31/9999 9:59:59 PM, Unspecified] => 12/31/9999 9:59:59 PM +08:00
 |--[12/29/9999 11:59:59 PM, Unspecified] => 12/29/9999 11:59:59 PM +08:00
 |--[12/31/9997 11:59:59 PM, Unspecified] => 12/31/9997 11:59:59 PM +08:00
 |--[7/9/2015 1:02:03 AM, Unspecified] => 7/9/2015 1:02:03 AM +08:00
 |--[7/9/2015 1:02:03 AM, Utc] => 7/9/2015 9:02:03 AM +08:00
 |--[7/9/2015 1:02:03 AM, Local] => 7/9/2015 6:02:03 AM +08:00
 |--[1/1/0001 1:01:01 AM, Unspecified] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 1:01:01 AM, Utc] => 1/1/0001 12:00:00 AM +00:00
 |--[1/1/0001 1:01:01 AM, Local] => 1/1/0001 12:00:00 AM +00:00
 |--[12/31/9999 11:01:01 PM, Unspecified] => 12/31/9999 11:01:01 PM +08:00
 |--[12/31/9999 11:01:01 PM, Utc] => 12/31/9999 11:59:59 PM +00:00
 |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 11:59:59 PM +00:00
```

I've compared it in winMerge and following lines are different:
```
> yours:  |--[7/9/2015 1:02:03 AM, Local] => 7/8/2015 10:02:03 AM -07:00
> mine:   |--[7/9/2015 1:02:03 AM, Local] => 7/8/2015 3:02:03 PM -07:00

> yours:  |--[7/9/2015 1:02:03 AM, Local] => 7/8/2015 10:02:03 PM +00:00
> mine:   |--[7/9/2015 1:02:03 AM, Local] => 7/8/2015 5:02:03 PM +00:00

> yours:  |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 8:01:01 PM +00:00
> mine:   |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 3:01:01 PM +00:00

> yours:  |--[7/9/2015 1:02:03 AM, Local] => 7/9/2015 6:02:03 AM +08:00
> mine:   |--[7/9/2015 1:02:03 AM, Local] => 7/9/2015 1:02:03 AM +08:00

> yours:  |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 11:59:59 PM +00:00
> mine:   |--[12/31/9999 11:01:01 PM, Local] => 12/31/9999 11:01:01 PM +08:00
```

**Comment by xuzhg** [comment]
@CrazyViper I think it's right. Because we are in different TimeZone, so, for the local kind of DateTime, we get the different output. 

If you haven't any concern, would you please update and rebase your PR? If you haven't time, I'd happy to do it for you. :smile:

**Comment by CrazyViper** [comment]
@xuzhg Yes, I think you're right.

Currently I have not time to do good test-cases to update my PR, so if you can, do this for me. Thanks!

**Comment by xuzhg** [comment]
@CrazyViper  I merged. If you have time, would you please try it out and let us know any issues. Thanks.

By the way, "pmuzyka" is your account?

**Comment by CrazyViper** [comment]
@xuzhg Thanks, I'll try it out later today.

Yes, 'pmuzyka' is my secondary e-mail. I've attached it to my account now.

**Comment by CrazyViper** [comment]
@xuzhg I've checked that code and I think that there're couple issues in it:

1. if (timeZone.BaseUtcOffset.Hours >= 0) can fail with offset="-00:30:00" (negative 30 minutes). In real timeZones there're some with offsets that have minutes, not only hours

2. Condition using timeZone.BaseUtcOffset will fail for DST offsets. That's why using  timeZone.GetUtcOffset(dateTime) is better.

**Comment by xuzhg** [comment]
@CrazyViper What we considered is the standard TimeZones. Is there a timezone with the above offset?

**Comment by congysu** [comment]
@xuzhg Please find in Wikipedia:
>a few are offset by 30 or 45 minutes (for example Newfoundland Standard Time is UTC -03:30 and Nepal Standard Time is UTC +05:45).
https://en.wikipedia.org/wiki/Time_zone

**Comment by xuzhg** [comment]
@CrazyViper I modified my codes as follows:

```C#
private static DateTimeOffset ConvertDateTimeToTimeOffset2(DateTime dateTime, TimeZoneInfo timeZone)
{
    TimeSpan utcOffset = timeZone.GetUtcOffset(dateTime);
    if (utcOffset >= TimeSpan.Zero)
    {
        if (dateTime <= DateTime.MinValue + utcOffset)
        {
            return DateTimeOffset.MinValue;
        }
    }
    else
    {
        if (dateTime >= DateTime.MaxValue + utcOffset)
        {
            return DateTimeOffset.MaxValue;
        }
    }

    if (dateTime.Kind == DateTimeKind.Local)
    {
        TimeZoneInfo localTimeZoneInfo = TimeZoneInfo.Local;
        TimeSpan localTimeSpan = localTimeZoneInfo.GetUtcOffset(dateTime);
        if (localTimeSpan < TimeSpan.Zero)
        {
            if (dateTime >= DateTime.MaxValue + localTimeSpan)
            {
                return DateTimeOffset.MaxValue;
            }
        }
        else
        {
            if (dateTime <= DateTime.MinValue + localTimeSpan)
            {
                return DateTimeOffset.MinValue;
            }
        }

        return TimeZoneInfo.ConvertTime(new DateTimeOffset(dateTime), timeZone);
    }

    if (dateTime.Kind == DateTimeKind.Utc)
    {
        return TimeZoneInfo.ConvertTime(new DateTimeOffset(dateTime), timeZone);
    }

    return new DateTimeOffset(dateTime, timeZone.GetUtcOffset(dateTime));
}
```

**Comment by xuzhg** [comment]
Closed. Please open a new issue if you find any further problem.

**Pull request closed without merging**